metals/src/main/scala/scala/meta/internal/metals/ScalafixProvider.scala

```diff
@@ -35,52 +39,76 @@ case class ScalafixProvider(
       scalaVersion: ScalaVersion,
       scalaBinaryVersion: ScalaBinaryVersion,
       classpath: JList[Path]
-  ): List[l.TextEdit] = {
+  ): Future[List[l.TextEdit]] = {
     val fileInput = file.toInput
     val unsavedFile = file.toInputFromBuffers(buffers)
     if (isUnsaved(unsavedFile.text, fileInput.text)) {
       scribe.info(s"Organize imports requires saving the file first")
       languageClient.showMessage(
         MessageType.Warning,
-        s"Save ${file.toNIO.getFileName} before organizing imports"
+        s"Save ${file.toNIO.getFileName} to compile it before organizing imports"
       )
-      Nil
+      Future.successful(Nil)
     } else {
-      val scalafixConfPath = workspace.resolve(scalafixConfigPath)
-      val scalafixConf =
-        if (scalafixConfPath.isFile) Some(scalafixConfPath.toNIO)
-        else None
-      val scalafixEvaluation = for {
-        api <- getScalafix(scalaBinaryVersion)
-        urlClassLoaderWithExternalRule <- getRuleClassLoader(
-          scalaBinaryVersion,
-          api.getClass.getClassLoader
-        )
-      } yield {
-        val scalacOption =
-          if (scalaBinaryVersion == "2.13") "-Wunused:imports"
-          else "-Ywarn-unused-import"
-
-        api
-          .newArguments()
-          .withScalaVersion(scalaVersion)
-          .withClasspath(classpath)
-          .withToolClasspath(urlClassLoaderWithExternalRule)
-          .withConfig(scalafixConf.asJava)
-          .withRules(List(organizeImportRuleName).asJava)
-          .withPaths(List(file.toNIO).asJava)
-          .withSourceroot(workspace.toNIO)
-          .withScalacOptions(Collections.singletonList(scalacOption))
-          .evaluate()
-      }
+      compilations.compilationFinished(file).map { _ =>
+        val scalafixConfPath = workspace.resolve(scalafixConfigPath)
+        val scalafixConf =
+          if (scalafixConfPath.isFile) Some(scalafixConfPath.toNIO)
+          else None
+        val scalafixEvaluation = for {
+          api <- getScalafix(scalaBinaryVersion)
```

Comment:
I would change this for comprehension from option to Try to get errors when failing downloading scalafix or the rule. It will also be easier to add error message on other part.
```
  private def getScalafix(
      scalaBinaryVersion: ScalaBinaryVersion
  ): Try[Scalafix] = {
    scalafixCache
      .get(scalaBinaryVersion)
      .toTry
      .orElse(
        statusBar.trackBlockingTask("Downloading scalafix") {
          Try(Scalafix.fetchAndClassloadInstance(scalaBinaryVersion))
            .map { api =>
              scalafixCache.update(scalaBinaryVersion, api)
              api
            }
        }
      )
  }

  private def getRuleClassLoader(
      scalaBinaryVersion: ScalaBinaryVersion,
      scalafixClassLoader: ClassLoader
  ): Try[URLClassLoader] = {
    val pathsOpt = organizeImportRuleCache
      .get(scalaBinaryVersion)
      .toTry
      .orElse(
        statusBar.trackBlockingTask("Downloading organize import rule") {
          Try(Embedded.organizeImportRule(scalaBinaryVersion))
            .map { paths =>
              organizeImportRuleCache.update(scalaBinaryVersion, paths)
              paths
            }
        }
      )
    pathsOpt.map(paths =>
      Embedded.toClassLoader(
        Classpath(paths.map(AbsolutePath(_))),
        scalafixClassLoader
      )
    )
  }
// to add with the rest of extension methods we have on Option.
  implicit class OptionExtension[A](val in: Option[A]) extends AnyVal {
    def toTry(e: => Throwable): Try[A] = in match {
      case Some(v) => Success(v)
      case None => Failure(e)
    }

    def toTry = toTry(new Exception("None.get "))
  
```

This way we will cover this error.

Reply:
Thanks, I changed it a bit, but stayed why Try as you suggested.